### PR TITLE
Pip: Normalize package name

### PR DESCRIPTION
**Pull request by tsteenbe**
In Python package names are case-insensitive, and uses of hyphens and underscores are considered to be equivalent.
This may lead to cases where the same dependency appears twice in ORT scan results with different cased ids.

For example, this happens if say in transitive dependency on https://pypi.org/project/Flask has been declared as either
[Flask][1] or [flask][2].

[1]: https://github.com/corydolphin/flask-cors/blob/3.0.10/requirements.txt
[2]: https://github.com/rycus86/prometheus_flask_exporter/blob/0.18.1/requirements.txt

**This PR is a breaking change for all ORT users analyzing Python projects**

After this PR is merged Python packages may get a different package id assigned by ORT in the scan results. If you have Python packages defined in your [curations.yml](https://github.com/oss-review-toolkit/ort/blob/master/docs/config-file-curations-yml.md) or [package configuration](https://github.com/oss-review-toolkit/ort/blob/master/docs/config-file-package-configuration-yml.md) then you will need to update all Python package ids to be lowercase and replace any `.` or `_` with a `-`.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 1]
```diff
@@ -553,7 +553,11 @@ class Pip(
                 id = Identifier(
                     type = "PyPI",
                     namespace = "",
-                    name = dependency["package_name"].textValue(),
+                    // Normalize all PyPI package names to be lowercase and hyphenated as per PEP 426:
+                    // "All comparisons of distribution names MUST be case insensitive,"
+                    // and MUST consider hyphens and underscores to be equivalent.
+                    // https://www.python.org/dev/peps/pep-0426/#name
+                    name = dependency["package_name"].textValue().toLowerCase().replace('_', '-').replace('.', '-'),
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/Pip.kt:560]: There is a look-up in using the identifier via a `Map` in line 355 which is case sensitive.
Thus, I believe the package name assigned by `parsePipShowOutput()` needs to also be normalized in order to not break the lookup.

**Review by fviernau — CHANGES_REQUESTED** [review]

**Review by fviernau — APPROVED** [review]

**Review by fviernau — APPROVED** [review]

**Comment by sschuberth** [comment]
This is still a draft. Is it ready for review yet?

**Comment by tsteenbe** [comment]
@sschuberth  Test are failing and I can't get my local python setup to work so I can locally run the fun test with `./gradlew analyzer:funTest --tests PipFunTest`. The tests are likely failing due to small diff in expected files but still have not been able to debug it.

**Comment by sschuberth** [comment]
@tsteenbe can you please still rebase to resolve the conflicts?

> In Python package names are case-insensitive, and uses of hyphens and underscores are considered to be equivalent.
> This may lead to cases where the same dependency appears twice in ORT scan results with different cased ids.

And could you please add a reference link for this statement to the commit message?

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 2]
```diff
@@ -96,7 +96,7 @@ object VirtualEnv : CommandLineTool {
     // Allow to use versions that are known to work. Note that virtualenv bundles a version of pip. Exclude version 20.3
     // for now as is comes with a new dependency resolver that we yet need to test, see
     // http://pyfound.blogspot.com/2020/11/pip-20-3-new-resolver.html.
-    override fun getVersionRequirement(): Requirement = Requirement.buildIvy("[15.1,20.3[")
+    override fun getVersionRequirement(): Requirement = Requirement.buildIvy("[15.1,20.5[")
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pip.kt:99]: Note: This should not be done without also taking the above code comment into account (i.e. test pip's new dependency resolver), and in any case it should be done in separate commit. In the context of this PR, I'd propose to simply drop this change.

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 3]
```diff
@@ -751,3 +751,9 @@ private fun Package.enrichWith(other: Package?): Package =
     } else {
         this
     }
+
+// Normalize all PyPI package names to be lowercase and hyphenated as per PEP 426:
+// "All comparisons of distribution names MUST be case insensitive,
+// and MUST consider hyphens and underscores to be equivalent".
+// https://www.python.org/dev/peps/pep-0426/#name  
+private fun String.normalizePackageName(): String = toLowerCase().replace('_', '-').replace('.', '-')
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pip.kt:759]: The docs say nothing about considering also `.` and `-` to be equivalent, so where does that come from?

Nit: The comment should be turned into KDoc with a `/** ... */`-style multi-line comment.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 2 above)
Comment (tsteenbe) [reply analyzer/src/main/kotlin/managers/Pip.kt]: Accidentally committed this, reverted it - my local Python version is more modern version so I needed this to be able to run the functional testst

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 3 above)
Comment (tsteenbe) [reply analyzer/src/main/kotlin/managers/Pip.kt:768]: Update the comment to included PEP 503 which explains this

**Review by sschuberth — APPROVED** [review]

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pip.kt:768]: Actually, PEP 503 suggests to use this code:

```
def normalize(name):
    return re.sub(r"[-_.]+", "-", name).lower()
```

Which means that also multiple consecutive `-` (or the other chars) will be collapsed to a single `-`, which is not something your current implementation does. Should we change to use a Regex, too?

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pip.kt:768]: ... and do the `toLowerCase()` as the last step, too, as the string might be shorter then 😁

Comment (tsteenbe) [reply analyzer/src/main/kotlin/managers/Pip.kt:768]: @sschuberth  If you can provide me with the Kotlin code to multiple consecutive `-` (or the other chars), I am happy to make the update.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pip.kt:768]: It's mostly just copy & paste of the Python code, and you can easily play around at the Kotlin playground, see

https://pl.kotl.in/bIoKb45gN

**Review by sschuberth — APPROVED** [review]

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 3 above)
Comment (tsteenbe) [reply analyzer/src/main/kotlin/managers/Pip.kt]: Did not know about Kotlin playground thanks for showing me, fixed.
